libs/oauth-key-service/src/main/java/uk/gov/di/ipv/core/library/oauthkeyservice/OAuthKeyService.java

```diff
@@ -51,64 +65,114 @@ public RSAKey getEncryptionKey(OauthCriConfig criConfig) throws ParseException {
         var jwksUrl = criConfig.getJwksUrl();
         var keyFromConfig = criConfig.getParsedEncryptionKey();
 
-        if (jwksUrl != null) {
-            if (cachedOAuthEncryptionKeys.containsKey(jwksUrl.toString())
-                    && !cachedOAuthEncryptionKeys.get(jwksUrl.toString()).isExpired()) {
-                return cachedOAuthEncryptionKeys.get(jwksUrl.toString()).key();
-            }
+        return jwksUrl == null
+                ? keyFromConfig
+                : getCachedJWKSet(jwksUrl).filter(ENC_USE_MATCHER).getKeys().stream()
+                        .findFirst()
+                        .map(JWK::toRSAKey)
+                        .orElseGet(
+                                () -> {
+                                    LOGGER.warn(
+                                            LogHelper.buildLogMessage(
+                                                            "No encryption key found, returning key from config")
+                                                    .with(
+                                                            LOG_CRI_ISSUER.getFieldName(),
+                                                            criConfig.getComponentId())
+                                                    .with(LOG_JWKS_URL.getFieldName(), jwksUrl));
+                                    return keyFromConfig;
+                                });
+    }
 
-            var keys = getKeysFromJwksEndpoint(jwksUrl);
+    public ECKey getClientSigningKey(String clientId, JWSHeader jwsHeader) throws ParseException {
```

Comment:
(Very minor - do you think it's more appropriate to pass the keyId here instead of the entire header?)

Reply:
I think not - we're extracting the `kid` to find the right key, but we could go further. You can create a matching object directly from a header, which could use things like the algorithm and the use to refine the JWKSet. I think it'd be better to leave the service to decide what info it wants to use to find the key - will avoid having to update anything that consumes this method if we change the logic in future.